readme.md
# How to Use?
This is a project for the 2023 study of the Sensorium research project. \\
All copyrights belong to Sanghoon Lee, Ji-yeong Chae. \\
For questions, please email leesh2913@dgist.ac.kr.

# preprocess.py
process raw file to **int** csv file   
csv file is saved at **data/x_data_ .csv** and **data/y_data_ .csv**   
## pre_process_00
save only frequency (0 ~ 5120hz data)   
data name is **data/x_data_00.csv**, **data/y_data_00.csv**

## pre_process_01
frequency (0 ~ 5120hz data) + Pipe info (p pd s type data)
data name is **data/x_data_01.csv**, **data/y_data_01.csv**

# train.py
## 1. import train.py
```
import train
```
## 2. Initialize train class   
```
trainer = train.train(name)
``` 
name is the file name of **/data folder**
example: "00", "01"   

## 3. search_parameter(model_name, rand_seed, parameter, name)   
```
params = {'n_neighbors': [5, 15, 30, 100], 'weights':  ['uniform', 'distance'],
          'algorithm': ['auto', 'ball_tree', 'kd_tree', 'brute']}

trainer.search_parameter("KNN", 3, param, "First test")
```
- **model_name** is the name of ML methods   
    examples: "KNN", "MLP", "DNN", "GB", etc ...

- **rand_seed** is the seed of randome state
    examples: 0, 42, 1040, etc ...

- **parameter** is the dictionary of parameter set
    examples: 
    ```
    params = {
        'hidden_layer_sizes': [(50,), (100,), (150,)],
        'activation': ['tanh', 'relu', 'logistic'],
        'solver': ['adam'],
        'alpha': [0.001, 0.01, 0.1]
    }
    ```
    You can check the parameter settings in **parameter_list.txt**

- **name** is the file name that you want to save
    examples: 0234   
    If **model_name** is MLP and **name** is "All" then   
    the result of GridSearchCV is saved at **log/MLP_ALL.csv**

# Parameter_list.txt
Save the possible parameter list of any M.L methods.


history/history.txt
def pre_process_00(name):
    data = open('data/raw_data.csv')
    data = csv.reader(data)
    header = next(data)
    
    freq_data = np.array([])
    target_data = np.array([])
    
    for row in data:
        if len(freq_data) == 0:
            freq_data = np.append(freq_data, np.array(row[6:]), axis=0)
        else:
            freq_data = np.vstack([freq_data, np.array(row[6:])])
        if (row[4] == '0' or row[4] == '1'):
            target_data = np.append(target_data, np.array(0))
        if (row[4] == '2' or row[4] == '3' or row[4] == '4'):
            target_data = np.append(target_data, np.array(1))
        
    x_data = freq_data
    x_data = np.array(x_data, int)
    
    x_df = pd.DataFrame(x_data)
    x_name = 'data/x_data_' + name + '.csv' 
    x_df.to_csv(x_name, index=False)
    
    target_data = np.array(target_data, int)
    y_df = pd.DataFrame(target_data)
    y_name = 'data/y_data_' + name + '.csv' 
    y_df.to_csv(y_name, index=False)
    
def pre_process_01(name):
    data = open('data/raw_data.csv')
    data = csv.reader(data)
    header = next(data)
    
    freq_data = np.array([])
    pipe_data = np.array([])
    target_data = np.array([])
    
    for row in data:
        if len(freq_data) == 0:
            freq_data = np.append(freq_data, np.array(row[6:]), axis=0)
            pipe_data = np.append(
                pipe_data, np.array(row[1:4]), axis=0)
        else:
            freq_data = np.vstack([freq_data, np.array(row[6:])])
            pipe_data = np.vstack([pipe_data, np.array(row[1:4])])
        if (row[4] == '0' or row[4] == '1'):
            target_data = np.append(target_data, np.array(0))
        if (row[4] == '2' or row[4] == '3' or row[4] == '4'):
            target_data = np.append(target_data, np.array(1))
        
    x_data = np.concatenate([pipe_data, freq_data], axis=1)
    x_data = np.array(x_data, int)
    
    x_df = pd.DataFrame(x_data)
    x_name = 'data/x_data_' + name + '.csv' 
    x_df.to_csv(x_name, index=False)
    
    target_data = np.array(target_data, int)
    y_df = pd.DataFrame(target_data)
    y_name = 'data/y_data_' + name + '.csv' 
    y_df.to_csv(y_name, index=False)

def pre_process_02(name):
    data = open('data/raw_data.csv')
    data = csv.reader(data)
    header = next(data)
    
    freq_data = np.array([])
    pipe_data = np.array([])
    target_data = np.array([])
    
    for row in data:
        arr = []
        # pipe data: augmentation
        for i in np.array(row[1:4]):
                arr.extend([i for j in range(10)])
        if len(freq_data) == 0:
            freq_data = np.append(freq_data, np.array(row[6:]), axis=0)
            pipe_data = np.append(
                pipe_data, np.array(np.array(arr)), axis=0)
        else:
            freq_data = np.vstack([freq_data, np.array(row[6:])])
            pipe_data = np.vstack([pipe_data, np.array(arr)])
        if (row[4] == '0' or row[4] == '1'):
            target_data = np.append(target_data, np.array(0))
        if (row[4] == '2' or row[4] == '3' or row[4] == '4'):
            target_data = np.append(target_data, np.array(1))
        
    x_data = np.concatenate([pipe_data, freq_data], axis=1)
    x_data = np.array(x_data, int)
    
    x_df = pd.DataFrame(x_data)
    x_name = 'data/x_data_' + name + '.csv' 
    x_df.to_csv(x_name, index=False)
    
    target_data = np.array(target_data, int)
    y_df = pd.DataFrame(target_data)
    y_name = 'data/y_data_' + name + '.csv' 
    y_df.to_csv(y_name, index=False)

def pre_process_03(name):
    data = open('data/raw_data.csv')
    data = csv.reader(data)
    header = next(data)
    
    freq_data = np.array([])
    pipe_data = np.array([])
    target_data = np.array([])
    
    for row in data:
        if row[1] == '-1':
            continue
        
        arr = []
        # pipe data: augmentation
        for i in np.array(row[1:4]):
                arr.extend([i for j in range(10)])
        if len(freq_data) == 0:
            freq_data = np.append(freq_data, np.array(row[6:]), axis=0)
            pipe_data = np.append(
                pipe_data, np.array(np.array(arr)), axis=0)
        else:
            freq_data = np.vstack([freq_data, np.array(row[6:])])
            pipe_data = np.vstack([pipe_data, np.array(arr)])
        if (row[4] == '0' or row[4] == '1'):
            target_data = np.append(target_data, np.array(0))
        if (row[4] == '2' or row[4] == '3' or row[4] == '4'):
            target_data = np.append(target_data, np.array(1))
        
    x_data = np.concatenate([pipe_data, freq_data], axis=1)
    x_data = np.array(x_data, int)
    
    x_df = pd.DataFrame(x_data)
    x_name = 'data/x_data_' + name + '.csv' 
    x_df.to_csv(x_name, index=False)
    
    target_data = np.array(target_data, int)
    y_df = pd.DataFrame(target_data)
    y_name = 'data/y_data_' + name + '.csv' 
    y_df.to_csv(y_name, index=False)

def pre_process_04(name):
    data = open('data/raw_data.csv')
    data = csv.reader(data)
    header = next(data)
    
    freq_data = np.array([])
    pipe_data = np.array([])
    target_data = np.array([])
    
    onehot_encoder = OneHotEncoder(sparse = False)
    scaler = StandardScaler()
    
    for row in data:
        
        # Remove -1
        if row[1] == '-1':
            continue
        
        arr = []
        # pipe data: augmentation
        # for i in np.array(row[1:4]):
        #         arr.extend([i for j in range(10)])
        if len(freq_data) == 0:
            freq_data = np.append(freq_data, np.array(row[6:]), axis=0)
            pipe_data = np.append(
                pipe_data, np.array(np.array(row[1:4])), axis=0)
        else:
            freq_data = np.vstack([freq_data, np.array(row[6:])])
            pipe_data = np.vstack([pipe_data, np.array(row[1:4])])
        if (row[4] == '0' or row[4] == '1'):
            target_data = np.append(target_data, np.array(0))
        if (row[4] == '2' or row[4] == '3' or row[4] == '4'):

KNN_실험2.py
# coding: utf-8
from sklearn.datasets import load_breast_cancer
import numpy as np
import matplotlib.pyplot as plt
import csv
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
import pandas as pd
from sklearn import metrics
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import KFold
from sklearn.metrics import precision_score, recall_score, f1_score, accuracy_score

cancer = load_breast_cancer()

f = open('/home/jy/Desktop/엑셀 형식 - 계량기만.csv')  # 1
f2 = open('/home/jy/Desktop/계량기&ltype1 지운 형식.csv')  # 1
data = csv.reader(f)
data2 = csv.reader(f2)
header = next(data)
header2 = next(data2)

cancer.data = np.array([])
cancer.data2 = np.array([])
cancer.extraData = np.array([])
cancer.extraData2 = np.array([])
cancer.target = np.array([])
cancer.target2 = np.array([])

cancer.feature_names = np.array(header[6:])
cancer.feature_names2 = np.array(header2[6:])  # OK
cancer.feature_names = np.append(
    cancer.feature_names, header[1:4])
cancer.feature_names2 = np.append(
    cancer.feature_names2, header[1:4])
cancer.target_names = np.array([0, 1])
cancer.target_names2 = np.array([0, 1])  # OK

for row in data:  # 2
    # print(1)
    if len(cancer.data) == 0:
        cancer.data = np.append(cancer.data, np.array(row[6:]), axis=0)
        cancer.extraData = np.append(
            cancer.extraData, np.array(row[1:4]), axis=0)
    else:
        cancer.data = np.vstack([cancer.data, np.array(row[6:])])
        cancer.extraData = np.vstack([cancer.extraData, np.array(row[1:4])])
    if (row[4] == '0' or row[4] == '1'):
        cancer.target = np.append(cancer.target, np.array(0))
    if (row[4] == '2' or row[4] == '3' or row[4] == '4'):
        cancer.target = np.append(cancer.target, np.array(1))

for row in data2:  # 2
    # print(1)
    if len(cancer.data2) == 0:
        cancer.data2 = np.append(cancer.data2, np.array(row[6:]), axis=0)
        cancer.extraData2 = np.append(
            cancer.extraData2, np.array(row[1:4]), axis=0)
    else:
        cancer.data2 = np.vstack([cancer.data2, np.array(row[6:])])
        cancer.extraData2 = np.vstack([cancer.extraData2, np.array(row[1:4])])
    if (row[4] == '0' or row[4] == '1'):
        cancer.target2 = np.append(cancer.target2, np.array(0))
    if (row[4] == '2' or row[4] == '3' or row[4] == '4'):
        cancer.target2 = np.append(cancer.target2, np.array(1))

cancer.data = np.concatenate([cancer.data, cancer.extraData], axis=1)
cancer.data2 = np.concatenate([cancer.data2, cancer.extraData2], axis=1)

f.close
f2.close

# 훈련/테스트 세트로 나누기
precision_micro = 0
precision_macro = 0
precision_weighted = 0
precision_binary = 0
recall_micro = 0
recall_macro = 0
recall_weighted = 0
recall_binary = 0
f1_micro = 0
f1_macro = 0
f1_weighted = 0
f1_binary = 0
accuracy_True = 0
accuracy_False = 0

params = {'n_neighbors': [5, 15, 30, 100], 'weights':  ['uniform', 'distance'],
          'algorithm': ['auto', 'ball_tree', 'kd_tree', 'brute']}
acc = []

# 훈련/테스트 세트로 나누기 (random_state 바꿔서)
for i in range(100):
    X_train, X_test, y_train, y_test = train_test_split(
        cancer.data, cancer.target, random_state=i)
    X_train2, X_test2, y_train2, y_test2 = train_test_split(
        cancer.data2, cancer.target2, random_state=i, test_size=149.25/410)
    knn = KNeighborsClassifier()
    gsknn = GridSearchCV(knn, params, cv=3,
                         scoring="f1_macro", verbose=1, refit=True)
    gsknn.fit(X_train, y_train)

    scores_df = pd.DataFrame(gsknn.cv_results_)
    scores_df = scores_df[["params", "mean_test_score", "rank_test_score"]]
    pd.set_option('display.max_colwidth', -1)
    acc.append(scores_df)
    # knn = KNeighborsClassifier(
    #    algorithm='auto', n_neighbors=100, weights='uniform')
    # knn.fit(X_train, y_train)
    # y_hat = knn.predict(X_test2)
    # matrix = metrics.confusion_matrix(y_test, y_hat)
    """
    parameter_array.append(gsknn.best_params_)
    estimator = gsknn.best_estimator_
    y_hat = estimator.predict(X_test2)
    accuracy_array.append(round(accuracy_score(y_test2, y_hat), 4))

    precision_micro += precision_score(y_test2, y_hat, average='micro')
    precision_macro += precision_score(y_test2, y_hat, average='macro')
    precision_weighted += precision_score(y_test2, y_hat, average='weighted')
    precision_binary += precision_score(y_test2,
                                        y_hat, average='binary', pos_label=0)
    recall_micro += recall_score(y_test2, y_hat, average='micro')
    recall_macro += recall_score(y_test2, y_hat, average='macro')
    recall_weighted += recall_score(y_test2, y_hat, average='weighted')
    recall_binary += recall_score(y_test2, y_hat,
                                  average='binary', pos_label=0)
    f1_micro += f1_score(y_test2, y_hat, average='micro')
    f1_macro += f1_score(y_test2, y_hat, average='macro')
    f1_weighted += f1_score(y_test2, y_hat, average='weighted')
    f1_binary += f1_score(y_test2, y_hat, average='binary', pos_label=0)
    accuracy_True += accuracy_score(y_test2, y_hat, normalize=True)
    accuracy_False += accuracy_score(y_test2, y_hat, normalize=False)
    # value_sum += forest.score(X_test, y_test)
    # print(matrix)
    # 최고 parameter -> 해당 matrix 뽑기
    # print("훈련 세트 정확도 : {:.3f}".format(Logreg.score(X_train, y_train)))
    # print("테스트 세트 정확도 : {:.3f}".format(Logreg.score(X_test, y_test)))
print(precision_micro / 100)
print(precision_macro / 100)
print(precision_weighted / 100)
print(precision_binary / 100)
print(recall_micro / 100)
print(recall_macro / 100)
print(recall_weighted / 100)
print(recall_binary / 100)
print(f1_micro / 100)
print(f1_macro / 100)
print(f1_weighted / 100)
print(f1_binary / 100)
print(accuracy_True / 100)
print(accuracy_False / 100)

# y_hat = forest.predict(X_test)
# matrix = metrics.confusion_matrix(y_test, y_hat)
# print(matrix)
# 얘로 최고 parameter 뽑고 나서 -> 해당 matrix 뽑기
scores_df = pd.DataFrame(gsknn.cv_results_)
scores_df = scores_df[["params", "mean_test_score",
                       "rank_test_score"]]
pd.set_option('display.max_colwidth', -1)
print(scores_df)
print("최적의 하이퍼 파라미터: ", gsknn.best_params_)
print("최고의 예측 정확도: ", gsknn.best_score_)
# print("훈련 세트 정확도 : {:.3f}".format(forest.score(X_train, y_train)))
# print("테스트 세트 정확도 : {:.3f}".format(forest.score(X_test, y_test)))

"""
df = pd.DataFrame(acc)
df.to_csv('KNN_실험2.csv')
